### PR TITLE
[NUI] Add ProcessorController

### PR DESCRIPTION
**Pull request by bshsqa**
This patch makes ProcessorController to use process of dali-core.
Added ProcessorCallback will be called at the end of event thread for every event tick.
If the callback become unnecessary, remove the callback from ProcessorController.

This patch includes to make LayoutController use ProcessorController.

This patch must be merged with https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-csharp-binder/+/260460/

Signed-off-by: seungho <sbsh.baek@samsung.com>

### CODE REVIEW

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        public void AddLayoutControllerProcessor(ProcessorInterface.ProcessorCallback layoutControllerProcessorCallback)
+        {
+            Interop.ProcessorController.AddLayoutControllerCallback(SwigCPtr, layoutControllerProcessorCallback);
+        }
+
+        /// <summary>
+        /// Remove Processor callback on this ProcessorController.
+        /// The removed callback will not be called anymore.
+        /// </summary>
+        public void RemoveProcessor(ProcessorInterface.ProcessorCallback processorCallback)
```
Comment (dongsug-song) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:83]: I have a question. 
1) if user do `AddProcesser(userCallback)` and do NOT `RemoveProcessor(userCallback)`, then `userCallback` is called continuously?
2) many `processorCallback` can be added? like this? `ProcessorController.Instance.AddLayoutControllerProcessor(callback1)`, ... ,  `ProcessorController.Instance.AddLayoutControllerProcessor(callback_n)` ?

Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs:83]: 1. Yes.
I wrote it on the comment of class above.
/// The added callbacks will be called every time of event tick.
/// If the callback is not be wanted to be called anymore, remove it.

2. Yes too. it is possible.

**Review by dongsug-song — APPROVED** [review]
looks good to me, but other members' reviews are required~

File: src/Tizen.NUI/src/internal/Layouting/LayoutController.cs [hunk 2]
```diff
@@ -107,10 +105,10 @@ public Animation GetCoreAnimation()
         /// </summary>
         internal void CreateProcessCallback()
         {
-            if (instance == null)
+            if (LayoutProcessorCallback == null)
             {
-                instance = new Callback(Process);
-                Interop.LayoutController.SetCallback(SwigCPtr, instance);
+                LayoutProcessorCallback = new ProcessorInterface.ProcessorCallback(Process);
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Layouting/LayoutController.cs:110]: I think layout's processor need to be considered the last processor of all NUI processor...
we can devide pre-process event and process event...

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    /// The added callbacks will be called every time of event tick.
+    /// If the callback is not be wanted to be called anymore, remove it.
+    /// </summary>
+    internal sealed class ProcessorController : Disposable
+    {
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        internal delegate void ProcessorCallback();
+
+        private EventHandler processorOnceEventHandler;
+        private ProcessorOnceEventCallbackType processorOnceEventCallback;
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:43]: this callback looks unnecessary.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    ///
+    /// The added callbacks will be called every time of event tick.
+    /// If the callback is not be wanted to be called anymore, remove it.
+    /// </summary>
+    internal sealed class ProcessorController : Disposable
+    {
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        internal delegate void ProcessorCallback();
+
+        private EventHandler processorOnceEventHandler;
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:42]: this private event handler looks unnecessary.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        private ProcessorOnceEventCallbackType processorOnceEventCallback;
+        private EventHandler processorEventHandler;
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:44]: this priavte EventHandler looks unnecessary.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        private ProcessorEventCallbackType processorEventCallback;
+        private EventHandler layoutProcessorEventHandler;
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:46]: this private EventHandler looks unnecessary.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        private LayoutProcessorEventCallbackType layoutProcessorEventCallback;
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:47]: this callback looks unnecessary.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    /// </summary>
+    internal sealed class ProcessorController : Disposable
+    {
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        internal delegate void ProcessorCallback();
+
+        private EventHandler processorOnceEventHandler;
+        private ProcessorOnceEventCallbackType processorOnceEventCallback;
+        private EventHandler processorEventHandler;
+        private ProcessorEventCallbackType processorEventCallback;
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:45]: this callback looks unnecessary.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        private delegate void ProcessorEventCallbackType();
+
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        private delegate void LayoutProcessorEventCallbackType();
+
+        private ProcessorCallback callback = null;
+
+        public event EventHandler ProcessorOnceEvent
+        {
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:61]: no need to implement add/remove for simple event.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                processorOnceEventHandler -= value;
+
+                if (processorOnceEventHandler == null)
+                {
+                    processorOnceEventCallback = null;
+                }
+            }
+        }
+
+        public event EventHandler ProcessorEvent
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:83]: no need to reimplement add/remove for simple event.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+                processorEventHandler -= value;
+
+                if (processorEventHandler == null)
+                {
+                    processorEventCallback = null;
+                }
+            }
+        }
+
+        public event EventHandler LayoutProcessorEvent
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:106]: no need to reimplement add/remove for simple event.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+                return instance;
+            }
+        }
+
+        public void Process()
+        {
+            processorOnceEventHandler?.Invoke(this, null);
+            processorOnceEventHandler = null;
+            processorEventHandler?.Invoke(this, null);
+            layoutProcessorEventHandler?.Invoke(this, null);
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:158]: we can call
ProcessorOnceEvent?.Invoke(this, null);
ProcessorOnceEvent = null;
ProcessorEvent?.Invoke(this, null);
LayoutProcessorEvent?.Invoke(this, null);

directly.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+            processorOnceEventHandler = null;
+            processorEventHandler = null;
+            layoutProcessorEventHandler = null;
+            GC.SuppressFinalize(this);
+
+            base.Dispose(type);
+        }
+
+        // Callback for View OnProcessOnceEvent signal
+        private void OnProcessorOnceEvent()
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:177]: unnecessary event callback.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        {
+            if (processorOnceEventHandler != null)
+            {
+                processorOnceEventHandler(this, null);
+            }
+        }
+
+        // Callback for View OnProcessEvent signal
+        private void OnProcessorEvent()
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:186]: unnecessary event callback.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        {
+            if (processorEventHandler != null)
+            {
+                processorEventHandler(this, null);
+            }
+        }
+
+        // Callback for View OnLayoutProcessEvent signal
+        private void OnLayoutProcessorEvent()
```
Comment (everLEEst) [inline src/Tizen.NUI/src/internal/Common/ProcessorController.cs:195]: unnecessary event callback.

**Comment by bshsqa** [comment]
@everLEEst 
Thank you I fixed them.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+                }
+                return instance;
+            }
+        }
+
+        /// <summary>
+        /// Add Processor callback on this ProcessorController.
+        /// The input processorCallback will be called end of this event tick.
+        /// </summary>
+        public void AddProcessor(ProcessorInterface.ProcessorCallback processorCallback)
```
[the remark this answers is not in the document]
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I fixed.

File: src/Tizen.NUI/src/internal/Layouting/LayoutController.cs (hunk 2 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Layouting/LayoutController.cs]: Currently we don't need to post process of layouting.
I think we can add post process when we need. thank you.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 3 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 4 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 5 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 6 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 7 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 8 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 9 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.
I've learned how to use event.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 10 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 11 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 13 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 14 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/internal/Common/ProcessorController.cs (hunk 15 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/internal/Common/ProcessorController.cs]: Thank you I removed it!.

File: src/Tizen.NUI/src/public/BaseComponents/ViewEvent.cs [hunk 17]
```diff
@@ -319,7 +319,6 @@ private void SendViewAddedEventToWindow(IntPtr data)
                 {
                     this.HoveredSignal().Disconnect(hoverEventCallback);
                 }
-
```
Comment (everLEEst) [inline src/Tizen.NUI/src/public/BaseComponents/ViewEvent.cs:322]: can you remove this change? this file must not be included this patch.

**Review by everLEEst — COMMENTED** [review]
LGTM, but unnecessary changes are included.
could you please check?

File: src/Tizen.NUI/src/public/BaseComponents/ViewEvent.cs (hunk 17 above)
Comment (bshsqa) [reply src/Tizen.NUI/src/public/BaseComponents/ViewEvent.cs]: Thank you I fixed.

**Review by everLEEst — APPROVED** [review]
LGTM.

**Comment by everLEEst** [comment]
it would be better to merge this patch and dali-csharp-binder after integration done.

**Pull request merged**